tools/codegen/api/structured.py

```diff
@@ -0,0 +1,94 @@
+from tools.codegen.model import *
+
+from tools.codegen.api.types import *
+import tools.codegen.api.cpp as cpp
+
+from typing import Union, List
+
+# This file describes the translation of JIT schema to the structured functions API.
+# This is similar to native API, but a number of historical problems with native
+# API have been fixed.
+
+def name(func: FunctionSchema) -> str:
+    name = str(func.name.name)
+    if func.name.overload_name:
+        name += f'_{func.name.overload_name}'
+    return name
+
+# Translation of types occuring in JIT arguments to a C++ argument type.
+# NB: For now, mutable doesn't do anything; but it could if we make
+# some more nominal types
+def argumenttype_type(t: Type, *, mutable: bool, binds: ArgName) -> CType:
+    # If it's a value type, do the value type translation
+    r = cpp.valuetype_type(t, binds=binds)
+    if r is not None:
+        return r
+
+    if isinstance(t, BaseType):
+        if t.name == BaseTy.Tensor:
+            return ConstRefCType(BaseCType('Tensor', binds))
+        else:
+            raise AssertionError(f"base type should have been value type {t}")
+    elif isinstance(t, OptionalType):
+        if t.elem == BaseType(BaseTy.Tensor):
+            raise AssertionError(
+                "optional tensor not supported by structured yet; to implement this "
+                "add OptionalTensor c.f. https://github.com/pytorch/pytorch/issues/51456"
+            )
+        elem = argumenttype_type(t.elem, mutable=mutable, binds=binds)
+        return OptionalCType(elem)
+    elif isinstance(t, ListType):
+        if t.elem == BaseType(BaseTy.Tensor):
+            raise AssertionError(
+                "list of tensor not supported by structured yet; to implement this "
+                "resolve torch::List issue, see "
+                "https://fb.workplace.com/groups/894363187646754/permalink/1149276442155426"
+            )
+        # TODO: delete these special cases; see tools.codegen.api.cpp--these
+        # must be changed in tandem, but there are problems; see
+        # https://github.com/pytorch/pytorch/pull/51485
```

Comment:
I don't think utility is necessarily the right way to fix this. In an ideal world, we would canonicalize all C++ types (e.g., expand type aliases, resolve namespaces) before doing conversions so it would be self evident that `IntArrayRef` is the same thing as `ArrayRef<int64_t>`. With this canonicalization, it wouldn't be necessary to have the special cases at all (ergo no utility). But for unfortunate reasons (people directly matching on the C++ type; see the pull request) this can't be done right now, and I don't really want to block this patch set on it. But I kind of want this to still be potentially painful to update so that if it is painful again in the future someone will get around to actually fixing this.

Reply:
I guess... my worry is that having two similar but not identical piles of conversions in different parts of the codebase will just look entropic (if it's noticed at all) and invite fixing things by hacking, e.g. a new conversion needs to be added but no structured kernels use it yet so don't bother adding it there, that kind of thing. Not a hill to die on though.

Reply:
I agree, it absolutely is entropic. I am mostly suggesting that if we are worried about the entropy, it is better to invest time in fixing https://github.com/pytorch/pytorch/pull/51485